File: OCP/Solution/MarsRoverKata/MissionManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using MarsRoverKata.Commands;

namespace MarsRoverKata
{
    public class MissionManager
    {
        public Rover Rover;
        public Mars Planet; 
        private Commander commander;
        
        public MissionManager(Rover rover)
        {
            Rover = rover;
            Planet = rover.Mars;
            commander = new Commander();
        }

        public string AcceptCommands(String commandString)
        {
            commander.Commands = new List<ICommand>();

            var commands = commandString.ToLower().ToCharArray();
            bool success = true;
            foreach (Char command in commands)
            {
                success = AcceptCommand(success, command);
            }

            return success ? String.Empty : "Some invalid commands were found!!!";
        }

        private bool AcceptCommand(bool success, Char command)
        {
            if (command.Equals('f'))
                commander.Accept(new ForwardCommand(Rover));
            else if (command.Equals('b'))
                commander.Accept(new BackwardCommand(Rover));
            else if (command.Equals('r'))
                commander.Accept(new TurnRightCommand(Rover));
            else if (command.Equals('l'))
                commander.Accept(new TurnLeftCommand(Rover));
            else if (command.Equals('m'))
                commander.Accept(new FireMissileCommand(Rover));
            else if (command.Equals('g'))
                commander.Accept(new FireMortarCommand(Rover));
            else
                success = false;
            return success;
        }

        public string ExecuteMission()
        {
            bool success = commander.ExecuteCommands();

            return success ? String.Empty : "An error occured while executing commands";
        }
    }
}

File: LSP+ISP/Java_Exercise/MarsRoverKata/src/main/java/MarsRoverKata/MissionManager.java
package MarsRoverKata;

import MarsRoverKata.Commands.*;
import org.apache.commons.lang3.StringUtils;

import java.beans.EventHandler;
import java.util.ArrayList;

public class MissionManager {
    private Rover rover;

    public Rover getRover() {
        return rover;
    }

    public Mars planet;

    public Mars getPlanet() {
        return planet;
    }

    private final Commander _commander;

    public MissionManager(Rover rover) {
        this.rover = rover;
        planet = rover.getMars();
        _commander = new Commander();
        //MissionManager MissionManager = EventHandler.create(this.getClass(), this, "updateAliens", "");
        EventHandler eventHandler = new EventHandler(this, "updateAliens", null, null);
        _commander.setEventHandler(eventHandler);
    }

    //TODO: EventHandler implementation
    public void updateAliens() {
        getPlanet().updateAliens();
    }

    public String acceptCommands(String commandString) {
        _commander.setCommands(new ArrayList<ICommand>());

        char[] commands = commandString.toLowerCase().toCharArray();
        boolean success = true;
        for (char command : commands) {
            success = acceptCommand(success, command);
        }

        return success ? StringUtils.EMPTY : "Some invalid commands were found!!!";
    }

    private boolean acceptCommand(boolean success, char command) {
        if (command == 'f')
            _commander.accept(new ForwardCommand(rover));
        else if (command == 'b')
            _commander.accept(new BackwardCommand(rover));
        else if (command == 'r')
            _commander.accept(new TurnRightCommand(rover));
        else if (command == 'l')
            _commander.accept(new TurnLeftCommand(rover));
        else if (command == 'm')
            _commander.accept(new FireMissileCommand(rover));
        else if (command == 'g')
            _commander.accept(new FireMortarCommand(rover));
        else
            success = false;
        return success;
    }

    public String executeMission() {
        boolean success = _commander.executeCommands();
        return success ? StringUtils.EMPTY : "An error occured while executing commands";
    }

    public void addObstacle(int x, int y, String type) {
        Point location = new Point(x, y);
        IObstacle obstacle = createObstacle(location, type);
        getPlanet().addObstacle(obstacle);
    }

    private IObstacle createObstacle(Point location, String type) {
        if (type.equalsIgnoreCase("Alien")) {
            return new Alien(getPlanet(), location);
        }
        return new Obstacle(location);
    }
}

